[ENT-125] Dynamically load course modes from enrollment API

haikuginger opened the pull request:
**Description:**

Previously, the course mode box for enrolling users from the Enterprise app was limited to three "blessed" modes - audit, verified, and professional. However, it became clear that this was insufficient.

This PR modifies the view to dynamically load all available enrollment modes for a course directly from the Enrollment API, and presents them with the custom names given to them, rather than label them by internal value.

In addition, the other three valid modes (no-id-professional, credit, and honor) have been added to the allowable options, meaning that in cases where Javascript is disabled, all valid modes can be chosen correctly.

**Screenshots:**

No course ID entered:

![screen shot 2017-01-04 at 9 22 26 pm](https://cloud.githubusercontent.com/assets/7773758/21667301/b32f9480-d2c5-11e6-8ff4-157242817f34.png)

Valid course ID entered:

![screen shot 2017-01-04 at 9 23 42 pm](https://cloud.githubusercontent.com/assets/7773758/21667300/b32f3ee0-d2c5-11e6-9789-044353e83e2a.png)

Invalid course ID entered:

![screen shot 2017-01-04 at 9 24 05 pm](https://cloud.githubusercontent.com/assets/7773758/21667302/b330cd5a-d2c5-11e6-92e8-22a093ecdeea.png)

Javascript disabled:

![screen shot 2017-01-04 at 9 25 52 pm](https://cloud.githubusercontent.com/assets/7773758/21667299/b32b7ecc-d2c5-11e6-8bd6-793beba5f882.png)


**JIRA:** https://openedx.atlassian.net/browse/ENT-125

**Dependencies:** None

**Merge deadline:** 2017-01-17

**Sandbox**: https://ent91.opencraft.hosting (built-in `staff` user)

**Testing instructions:**

1. Create a couple custom course modes for a given course; use distinctive names.
2. Navigate to the Manage Learners view of the Enterprise Admin.
    * Observe that the course mode dropdown is disabled, and is filled with a prompt to enter a valid course ID.
2. Enter the course ID of the course you've chosen and tab away from the course box.
    * Observe that the course mode dropdown is populated with the course modes you created.
3. Select one of the modes and submit the form.
    * Observe that although the form has errors, the course mode dropdown is returned to the same state and position it was in prior to submitting the form.
4. Fill in the email address of an existing user not yet linked with this EnterpriseCustomer, and submit the form again.
    * Observe that the user is now enrolled in the given course, with the given course mode you chose from the dropdown.
3. Enter an invalid course ID in the course ID box.
    * Observe that the course mode box returns to its original, disabled, state.
4. Disable Javascript and refresh the page.
    * Observe that although nothing is dynamically loaded, all six valid course mode options are presented for you to select from.

**Reviewers:**
- [x] @e-kolpakov  
- [x] EdX TBD

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

1. I have some concerns about a11y with this PR. Multiple browsers (I tested Safari and Firefox) appear to choose tab order prior to calling the `.blur()` event on the control that's being departed. As a result, if the mode drop-down is disabled prior to tabbing away from the course ID box, then the focus will go to the submit button, rather than the dropdown, even if the AJAX call will enable the dropdown. There are a couple ways around this; none pretty:
    * Whenever course mode items are successfully loaded, set the focus to the dropdown. This is ugly because it prevents the user from choosing whatever focus item they like.
    * Don't actually disable the dropdown when there aren't valid options. This isn't ideal, but doesn't cause any harm, because when it's in such a state, there's only one option to choose from anyway.

edx-webhook commented:
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=26&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

e-kolpakov commented:
@haikuginger All looks good, except for a couple of things:

1. Version bump?
2. Changelog record?
3. (most important) That little JS file starts to grow - and there are no tests for it at all. Might be a good time to introduce some JS testing framework (i.e. Jasmine, qUnit, or anything you're familiar with).

mattdrayer commented:
@asadiqbal08 @saleem-latif @zubair-arbi  FYI on this PR --

haikuginger commented:
Rebased from f0cee6a.

@e-kolpakov, that 500 error should be fixed - an internal API client was trying to make an HTTPs connection to the LMS, which the hosts file listed as being at 127.0.0.1. Since we terminate HTTPs on the load balancer, this resulted in a failure of the API client to connect. I've also bumped the version and added a preliminary changelog entry.

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -1,5 +1,42 @@
+            });
+        }
+        function loadCourseModes(courseId){
+            $("#id_course").each(function(i, entry){
```
`i` ~ `index` ?

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -1,5 +1,42 @@
+            $("#id_course").each(function(i, entry){
+                courseId = entry.value
+                if (courseId === ''){
+                    disableMode()
```
nit: disableMode();

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -18,5 +55,20 @@ $(
                 })
             }
         });
+        $("#id_course").on('input', function(evt){
```
space around open bracket. 
nit: `evt` ~ `event`

e-kolpakov replied:
@asadiqbal08 @haikuginger technically it should be "space before opening curly-brace". :)

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -18,5 +55,20 @@ $(
+            */
+            evt.target.dirty = true;
+        });
+        $("#id_course").blur(function(evt){
```
space around open bracket. many places above . 
nit: good to have : `evt` ~ `event`

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -18,5 +55,20 @@ $(
+        });
+        $("#id_course").blur(function(evt){
+            var target = evt.target;
+            if (target.dirty){
```
space around open bracket. 

e.g. 
`if (target.dirty) {`

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -18,5 +55,20 @@ $(
+                target.dirty = false;
+            }
+        });
+        loadCourseModes()
```
what is the purpose of double calling `loadCourseModes()` ? 
nit: loadCourseModes()`;`

haikuginger replied:
It isn't being double-called. It's executed once on document ready, and as part of `$("#id_course").blur()`.

haikuginger commented:
@asadiqbal08, your comments have been addressed.

haikuginger commented:
Apologies to all for the ongoing emails as I try to get Travis to play nice with Jasmine.

haikuginger commented:
@e-kolpakov, JS test suite added.

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+                    }
+                ]
+            });
+            done()
```
done()`;`

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+                    },
+                    {
+                        name: 'booooo',
+                        slug: 'spooky ghost'
```
can we use some proper wording here too ?

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+                        slug: 'really squishy'
+                    },
+                    {
+                        name: 'booooo',
```
can we use some proper wording here instead of `booooo` ?

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+                course_modes: [
+                    {
+                        name: 'blah',
+                        slug: 'really squishy'
```
can we use some proper wording here too ?

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+            result = fillModeDropdown({
+                course_modes: [
+                    {
+                        name: 'blah',
```
can we use some proper wording here instead of `blah` ?

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+
+    describe("on page load", function(){
+        it("indicates that a valid course ID is required on page load.", function(){
+            loadPage();
```
instead of calling `loadPage();` in each function explicitly, you can put one place -- `beforeEach`

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+var enrollmentApiRoot = 'http://localhost/'
+var gettext = function(arg){
```
usually we are following to add space before opening braces in javascript files. better to follow the convention in all over places below too.

asadiqbal08 commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -1,22 +1,80 @@
+    });
+}
+
+function fillModeDropdown(data) {
```
can you add description that represent what this function `fillModeDropdown` is actually doing.

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+    it("makes the correct API call for course mode", function(){
+        loadPage();
+        $("#id_course").val("course1")
+        loadCourseModes()
```
loadCourseModes()`;`

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+
+    it("makes the correct API call for course mode", function(){
+        loadPage();
+        $("#id_course").val("course1")
```
`;`

asadiqbal08 commented on spec/javascripts/fixtures/learner_admin.html:
```diff
@@ -0,0 +1,6 @@
+    <option value="fake">Fake value</option>
+</select>
+
+<input id="id_course" value="">
```
line missing ?

asadiqbal08 commented on spec/javascripts/enterprise_admin_spec.js:
```diff
@@ -0,0 +1,71 @@
+            expect(dropdown).toContainElement("option[value='spooky ghost']")
+            expect(dropdown).toHaveText("blahbooooo")
+        });
+
```
extra line

e-kolpakov commented on spec/javascripts/fixtures/learner_admin.html:
```diff
@@ -0,0 +1,6 @@
+<select id="id_course_mode">
```
@haikuginger That's not even 10% of learner_admin HTML :) Suggestion: either rename it (to keep it a minimal amount of markup for course_mode tests) or extend to have a more or less actual HTML.

haikuginger commented:
@asadiqbal08, @e-kolpakov, updated to meet your requirements. Apologies about the consistent issues with Javascript formatting; I'm just not familiar with the style, so I can't "see" those issues quite yet.

e-kolpakov commented:
@haikuginger Suggestion: don't try to follow the JS convention - enforce it :) There's a bunch of JS linters, you might want to use jshint ([example commit of adding it](https://github.com/open-craft/xblock-group-project-v2/commit/bebb1304f2249e1991e5d86568296a48ddff8827)) or pick whatever you like.

haikuginger commented:
@mattdrayer, we've now got this running both Jasmine and jshint in CI; what can I do to help this get merged today to meet our sprint commitment?

mattdrayer commented:
I think we've gone through several rounds of revisions at this point, and I don't see anything important enough to warrant another set of edits.  Jenkins checks are green, so I say let the code flow -- 👍

asadiqbal08 commented:
Thanks @haikuginger 👍
